src/CoreWCF.BuildTools/src/OperationParameterInjectionGenerator.cs

```diff
@@ -0,0 +1,235 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics.SymbolStore;
+using System.Linq;
+using System.Text;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Text;
+
+namespace CoreWCF.BuildTools
+{
+    [Generator]
+    public class OperationParameterInjectionGenerator : ISourceGenerator
+    {
+        public void Execute(GeneratorExecutionContext context)
+        {
+            IEnumerable<SyntaxNode> allNodes = context.Compilation.SyntaxTrees
+                .SelectMany(x => x.GetRoot().DescendantNodes())
+                .ToImmutableArray();
+
+            var serviceContracts = FindServiceContracts(context, allNodes).ToImmutableArray();
+
+            foreach (var serviceImplAndContract in FindServiceImplAndContracts(context, allNodes, serviceContracts).ToImmutableArray())
+            {
+                var operationContracts = FindOperationContracts(context, serviceImplAndContract.contract).ToImmutableArray();
+
+                bool isErrorDiagnosticIssued = false;
+
+                foreach (var operationContract in operationContracts)
+                {
+                    var operationContractImplementation = serviceImplAndContract.service.FindImplementationForInterfaceMember(operationContract);
+
+                    if(operationContractImplementation == null)
+                    {
+                        var implementationCandidates = FindOperationContractImplementationCandidate(context, serviceImplAndContract.service, operationContract).ToImmutableArray();
+                        if(implementationCandidates.Length == 1)
+                        {
+                            if (serviceImplAndContract.isPartialClass)
+                            {
+                                GenerateOperationContractImplementation(context, serviceImplAndContract.service, serviceImplAndContract.contract, implementationCandidates[0], operationContract);
+                            }
+                            else if (!isErrorDiagnosticIssued)
+                            {
+                                context.ReportDiagnostic(DiagnosticDescriptors.ServicesShouldBePartialError(serviceImplAndContract.service.Name, serviceImplAndContract.contract.Name));
+                                isErrorDiagnosticIssued = true;
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        private static bool GetInstanceContextMode(GeneratorExecutionContext context, INamedTypeSymbol service)
+        {
```

Comment:
This doesn't get the InstanceContextMode, it just returns if it's a Single. I'm a little concerned that this is fragile as there are other ways to se the instance context mode. You can add a configure service delegate to your app which gets the ServiceBehaviorAttribute from the Description and modifies it. There is an internal property on InstanceContext called IsSingleton that we could make public. Currently it won't return the correct answer as it bases the answer on whether the IInstanceContextProvider is of type SingletonInstanceContextProvider and that's not the case any more. We could easily make that just an auto property which is set by ServiceBehavior.ApplyInstancing. Then you would just check InstanceContext.IsSingleton to decide whether to use a scope. If IsSingleton has a public setter too, then that would allow custom IInstanceContextProvider types to also set it and control this behavior.

Reply:
Thus i understand `InstanceContextMode` detection can only occur at runtime. I will adapt the generated code and expose the `IsSingleton` property as you suggested.

Reply:
> There is an internal property on InstanceContext called IsSingleton that we could make public. Currently it won't return the correct answer as it bases the answer on whether the IInstanceContextProvider is of type SingletonInstanceContextProvider and that's not the case any more. 

@mconnew I checked the code and i don't get why the `IsSingleton` is no longer returning the correct answer. So far in #473 we dit not introduce new `IInstanceContextProvider` implementation,
For `PerCall`/`PerSession` we adpated the `IInstanceProvider` implementation to pull `Service `instances from DI and expose a scoped `IServiceProvider` through `InstanceContext.Extensions`.
For `Single` we adapted the `ServiceBehaviorAttribute` to attach the `IServiceProvider` to the current `InstanceContext.Extensions`.
In both cases we did not alter `InstanceContext` instance lifetime,

Reply:
I exposed `IsSingleton` in #494

Reply:
I merged #494 so you can rebase and use that.